File: P/56_Merge_Intervals.java
//Array Sort (Time: O(NlogN) Space: O(N))
class Solution {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));
        
        LinkedList<int[]> list = new LinkedList();
        for(int[] edge: intervals) {
            if(list.isEmpty() || list.getLast()[1] < edge[0])
                list.add(edge);
            else
                list.getLast()[1] = Math.max(list.getLast()[1], edge[1]);
        }
        
        return list.toArray(new int[list.size()][]);
    }
}

//Tree Map (Time: O(NlogN) Space: O(N))
class Solution {
    public int[][] merge(int[][] intervals) {
        TreeMap<Integer, Integer> map = new TreeMap();
        
        for(int i = 0; i < intervals.length; i++)
            map.put(intervals[i][0], Math.max(map.getOrDefault(intervals[i][0], -1), intervals[i][1]));
        
        List<int[]> list = new ArrayList();
        while(!map.isEmpty()) {
            int start = map.firstKey();
            int end = map.get(start);
            map.remove(start);
            
            while(!map.isEmpty() && end >= map.firstKey()) {
                end = Math.max(end, map.get(map.firstKey()));
                map.remove(map.firstKey());
            }
            
            list.add(new int[] {start, end});
        }

        return list.toArray(new int[list.size()][]);
    }
}

File: 주영/Array/56. Merge Intervals.js
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function (intervals) {
  intervals.sort((a, b) => a[0] - b[0]);
  let arr = [];
  let prev = intervals[0];
  arr[0] = prev;

  for (let i = 0; i < intervals.length; i++) {
    let cur = intervals[i];

    if (cur[0] <= prev[1]) {
      prev[1] = Math.max(prev[1], cur[1]);
    } else {
      arr.push(cur);
      prev = cur;
    }
  }
  //console.log('arr: ', arr);
  return arr;
};

let intervals = [
  [2, 6],
  [1, 5],
  [9, 18],
  [19, 25],
];

merge(intervals);
